[c10d] explicitly abort communicators in destroy_process_group call

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119250

Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang committed:
[c10d] explicitly abort communicators in destroy_process_group call
Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

[ghstack-poisoned]

shuqiangzhang commented on collect_env.py:
Probably messed up with my repo, this file has nothing to do with my PR, will update the PR..

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1508,7 +1508,7 @@ void ProcessGroupNCCL::watchdogHandler() {
     dumpPipe.emplace(rank_);
   }
-  while (!done || !terminateProcessGroup_.load()) {
+  while (!done && !terminateProcessGroup_.load()) {
```
When terminateProcessGroup_ is true, watchdog will stop working (check workstatus, etc) immediately

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented:
cc @KaimingOuyang @eqy

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1522,7 +1522,13 @@ void ProcessGroupNCCL::watchdogHandler() {
+      // aborted, So cannot check exception based on them. But watchdog needs to
+      // finish the check for the works that have already been enqueued to
+      // workMetaList_
+      if (!terminateProcessGroup_.load()) {
```
@zdevito has learned that aborting a communicator causes the work object to 'complete'.  If we log this completion it can be a bit confusion.  We could handle this several ways

1) try to flush the work cleanup loop for some timeout before we actually do the aborts
1 seems not great imo, we don't really want to add more delays before doing the user-requested abort, nor can we be sure the delay would be fruitful (if collectives are hanging or taking longer than our timeout)

2) just do the abort and let the watchdog loop observe what it observes.  But add a log into the flight recorder for 'comm abort events', which should make it pretty clear from reading the flight log that any 'observed completions' after an abort are spurious
This ties into Zach's recent PR https://github.com/pytorch/pytorch/pull/119249

shuqiangzhang replied:
Discussed offline that we will roll out the change with knobs and will leave Flight recorder change to another PR

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1596,6 +1611,11 @@ def destroy_process_group(group: Optional[ProcessGroup] = None):
         pg._wait_for_pending_works()
 
     if group is None or group == GroupMember.WORLD:
+        # shutdown all backends in the order of pg names. shutting down in order because
```
is there any requirement that a child group is destroyed before the parent group that it was created from?

if we destroy in the reverse order that we create in, we could circumvent any issue like that.  cc @minsii

shuqiangzhang replied:
Good thing about PG name is that it is additive, e.g., first PG name is always PG0, second PG1, third PG2..... So we we want reverse order of How PGs are created, we just need to reverse sort the PG names

wconstab replied:
oh? i did not realize that. we do not let users name PG?

shuqiangzhang replied:
https://github.com/pytorch/pytorch/blob/dea15c9fdcd0d8cf09612ea2861541d1e743d5e5/torch/distributed/distributed_c10d.py#L3709

shuqiangzhang replied:
Changed the order to be the reverse of creation, e.g., PG3->PG2->PG1-PG0

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1327,6 +1327,21 @@ def _get_split_source(pg):
+def _shutdown_backend(pg):
+    """
+    Try to shut down the backend of a process group.
+    Currently, only ProcessGroupNCCL backend is supported.
```
wondering if it makes sense to plumb the shutdown call into all backends and let them be no-ops in the c++ layer, but i guess this is fine for now, and my idea is just more work

zdevito commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1327,6 +1327,21 @@ def _get_split_source(pg):
+        pass
+    if is_nccl_available() and isinstance(backend, ProcessGroupNCCL):
+        # explictly call shutdown to ensure that NCCL resources are released
+        backend._shutdown()
```
This is a new call for destroy_process_group, so it has the potential to introduce hangs in existing code. For instance, whereas previously we would clean up by calling ncclCommAbort in the destructor (which nominally shouldn't hang though it sometimes does), now we will explicitly be waiting for the shutdown call to complete, which _will_ hang if other ranks do not also call it. I think we need to make sure the completion of _shutdown is subject to the standard nccltimeout, and a timeout exception is raised if a rank doesn't also shutdown.

shuqiangzhang replied:
Yes, ideally ncclCommAbort should return either success or timeout, but instead it can hang forever. We should be able to add another layer of abstraction to implement this "either success or timeout"

eqy replied:
To clarify in case my understanding doesn't match what @zdevito is describing here: in NCCL 2.19.4+ we've observed that calling `ncclCommAbort` when other (intra-node) ranks don't also abort will also hang, and the response from NCCL is that while this was allowed for a number of versions, they are saying this is once again the expected behavior. In fewer words, I'm not sure this particular point (`_shutdown` can nominally hang while `ncclCommAbort` should not) will be true with current NCCL versions.

CC @KaimingOuyang to chime in

wconstab replied:
well keep in mind that users may use 2.17 for a while - unless we want to be really aggressive and mark that as unsupported.

so it seems like a desirable feature to have a timeout for ncclCommAbort, but since we can't rely on that being in older nccl verisons, we may have to build it in the pytorch layer for the time being.

KaimingOuyang replied:
I agree ncclCommAbort hanging is not pleasant behavior, unfortunately, this hang is limited by CUDA and we had no choice. For now, we have strict semantics that every active rank must call abort. However, it is possible to lift it by using CUMEM API and loosen the abort semantics in the future.

For old NCCL version, it does not hang since we directly free everything during abort, this can cause severe issue like segfault, but somehow, we got lucky and didn't meet the issue.

shuqiangzhang replied:
@KaimingOuyang "For old NCCL version, it does not hang since we directly free everything during abort", what is old NCCL version exactly? We can repro the hang consistently with version 2.17.

BTW, I got a working version of calling abort synchronously with a timeout in Pytorch layer, would update the PR soon after I clean up other stuffs.

KaimingOuyang replied:
Should be 2.14.x

shuqiangzhang replied:
Updated the PR to make abort() async with a timeout, throwing exception when timeout happens. Also add an env to gate the feature rollout.

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -987,6 +987,11 @@ def _is_barrier_after_init() -> int:
 
+def _abort_in_destroy_pg() -> bool:
+    # Environment variable to control whether to abort the communicators when users call destroy_process_group()
+    return os.getenv("TORCH_NCCL_ABORT_IN_DESTROY_PG") is not None
```
Need to be stricter by checking the value, will update in  the next version:
`os.getenv("TORCH_NCCL_ABORT_IN_DESTROY_PG") == "1" or os.getenv("TORCH_NCCL_ABORT_IN_DESTROY_PG").lower() == "true"`

wconstab replied:
yea, if we have the JK thing hooked up, the value will always be set by JK so this will be badly broken

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1115,57 +1162,71 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
   // communicators and signal the threads to exit. Joining on the threads could
   // potentially block and hence avoid it in this method.
   terminateProcessGroup_.store(true);
+  workMetaListCV_.notify_one();
```
so the intent here is to get the watchdog loop to wake up and shut itself down, right?

notifying it is good, but doesn't gaurantee it exits before our async abort kicks in.   its still a race whether ongoing works get updated by watchdog thread and observed as completed due to the abort right?  (do we care?)

shuqiangzhang replied:
Right, only after watchdog thread joined the destructor thread can we guarantee the exits. I think the watchdog only care the collectives before the shutdown call, but watchdog would still finish its while loop for already enqueued collectives (Though in normal cases, all collectives should complete before calling destroy_process_group). Even if there are non 0 left collectives, As we discussed, watchdog can 'record' when termination happens both in log and in flight recorder, which would be a separate PR

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1115,57 +1162,71 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
-  terminateProcessGroup_.store(true);
-  workMetaListCV_.notify_one();
 
+  if (!terminateProcessGroup_.load()) {
```
mental note- do you have tests for all 4 combinations of these 2 if's

shuqiangzhang replied:
Both true, covered by new test cases
Both false, covered by existing cases
We got covered 3 cases
terminateProcessGroup_ true, abortInDestroyProcessGroup_ false case, we can add a new test case to trigger the destructor
terminateProcessGroup_ false, but abortInDestroyProcessGroup_ is true , but we can add one to print the log

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1531,14 +1592,18 @@ void ProcessGroupNCCL::watchdogHandler() {
       if (work.exception()) {
         if (SHOULD_CLEAN_UP(asyncErrorHandling_)) {
-          // Abort work and corresponding communicators
-          work.abort();
```
not sure i understand why you removed work.abort but kept abort.  and whether you intended to clean up this codepath in this PR or later.

shuqiangzhang replied:
Yes, I wanted to clean up the work level abort eagerly, as work level abort does duplicate things as PG level abort, no need to call it twice. But if risky, I can leave it to later

wconstab replied:
ideally i would leave it as a separate PR and take the time to explain the reasoning.  (not that it has to wait to later, could just be in this stack of PRs)

shuqiangzhang replied:
Yes, the diff shown here is the 'outdated' version, the change has already been removed in the latest commit, marking as resolved

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1115,57 +1162,71 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
+              << "ProcessGroupNCCL heart beat monitor thread joined.";
   }
 #endif
+  if (onCompletionHookThread_.joinable()) {
```
oh we never joined this thread before eh?

well thats shoddy. one instinct is to add that join in a separate PR just in case we have to revert it and can keep the rest of this code landed

shuqiangzhang replied:
We called this join before, check line 1146 of this version of PR 
<img width="442" alt="image" src="https://github.com/pytorch/pytorch/assets/4711970/b3d4b243-e385-42bb-894b-f1ed39c99600">

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1056,6 +1058,51 @@ void ProcessGroupNCCL::waitForDumpOrTimeout(
   std::this_thread::sleep_until(wakeUpTime);
 }
 
+void ProcessGroupNCCL::waitForFutureOrTimeout(
```
nice helper. maybe in a followup PR you can replace my other wait functions with this? i think its better to keep that separate though.

shuqiangzhang replied:
Yes, this was on my list ^_^.

minsii commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1598,6 +1619,12 @@ def destroy_process_group(group: Optional[ProcessGroup] = None):
+        if _abort_in_destroy_pg():
+            # shutdown all backends in the order of pg names. shutting down in order because
+            # ncclCommAbort() was a 'collective' call in some versions of NCCL.
+            for pg_to_shutdown in sorted(_world.pg_names, key=lambda x: _world.pg_names[x], reverse=True):
```
I am not sure if we are completely deadlock free OK to rely on PG creation order, since nccl communicator may be created in different order, i.e., created when the first collective call is made with that pg.
```
pg0=create_group()
pg1=create_group()
...
allreduce(pg1)
// created pg1->comm
allreduce(pg0)
// created pg0->comm
...
```
Maybe better to based on actual nccl communicator creation, to be safe?

wconstab replied:
good point, i forgot about this!

shuqiangzhang replied:
Good point. but want to point out that this is 'best effort' ordering, it is just better than random order as of existing destruction order. it is not easy for python (e.g., without a c++ class level map) to track the actual creation order of ncclComms. 

Also Does it really matter to follow the exact reverse order of creation? I think as long as all ranks are terminating the PGs in the same order, we should be fine.

wconstab replied:
my theory about ordering (which i totally made up) was that a child group might need to be destroyed before the parent group it came from.  I don't know if order matters besides that.  (if you have 3 child groups from one parent, does it matter which order you destroy the children in? 

Agree that getting consistent ordering across ranks is a basic requirement and an improvement.

shuqiangzhang replied:
I considered this as resolved then? from unit tests of 2.19, the exact order within the same rank does not make a difference, but we can keep an eye on this, if it really matter, we can always address it later

shuqiangzhang replied:
Also, we need to consider if the children's comm is 'split' from the parent comm.  if not split, the destruction order might not matter. If children is split from parent, the real coms creation order must still be parent first (PG0), children later (e.g. PG1), which is the same as the python layer creation order, irrespective of which PG calls the collective first.  This is a requirement of ncclCommSplit (no outstanding op on parent comms: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommsplit)
https://github.com/pytorch/pytorch/blob/fa071a2e1b6617ba722d639747f1d885017b0e86/torch/csrc/distributed/c10d/NCCLUtils.hpp#L299

minsii replied:
I agree that we are good to destroy communicator in arbitrary order, as long as every rank follows the same order. I was thinking some potential circular deadlock case, where ranks create partially overlapped communicator in different order (e.g., FSDP PG + TP PG). But I guess it would still be fine, since we already force every rank to follow the same destroy order.

minsii commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -89,13 +89,22 @@ static std::vector<std::string> TORCH_NCCL_WAIT_TIMEOUT_DUMP_MILSEC = {
+// If yes, communicators will be aborted when destroy_process_group is called,
+// but not in destructor.
+static std::vector<std::string> TORCH_NCCL_ABORT_IN_DESTROY_PG = {
+    "TORCH_NCCL_ABORT_IN_DESTROY_PG"};
```
In ideal case, calling `ncclCommDestroy` (which is collective) is better than calling `ncclCommAbort` in normal termination. We can make the change in separate PR, but maybe better to avoid using "ABORT" in user facing control variable?

wconstab replied:
what would you propose for the env name?

shuqiangzhang replied:
Note that 'abort' is not naming after NCCL level ncclCommAbort, but PG level abort(), we can switch to ncclCommDestroy under the hood without changing user interfaces. 

Also note this ENV is here just for the roll out purpose, the plan is to delete it after the rollout

shuqiangzhang commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -1207,6 +1230,28 @@ def test_close_multi_pg_unordered(self):
+        self.assertEqual(backend.comm_split_count(), 2)
+        # shutdown all NCCL PGs in one shot
+        dist.destroy_process_group()
+
```
Good point, Will has similar coverage comment. added 3 more test cases. the warning print one is:
![image](https://github.com/pytorch/pytorch/assets/4711970/ce51bc2b-adec-4ea6-ae3b-e024f66df48e)

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -989,6 +989,12 @@ def _is_barrier_after_init() -> int:
     return int(os.getenv("TORCH_DIST_INIT_BARRIER", "0"))
 
 
+def _abort_in_destroy_pg() -> bool:
+    # Environment variable to control whether to abort the communicators when users call destroy_process_group()
+    env = os.getenv("TORCH_NCCL_ABORT_IN_DESTROY_PG", "0")
+    return env == "1" or env.lower() == "true"
+
+
```
Can this env ideally be avoided?
What are the pros and cons of defaulting to true?

kwen2501 replied:
The env name may have a mismatch with the layer this operates, i.e. we are operating at python PG level here but the env name has NCCL in it.

wconstab replied:
we need an env in order to have a way to do a smooth rollout

we can delete NCCL from the name but i think its better to keep it, since all of our PG envs are named `TORCH_NCCL`

shuqiangzhang replied:
I agree this env is not ideal/needed in the long term.  this is the common tradeoff of long term maintenance burden (lot of legacy knobs) vs short term rollout safety. I think the compromise of both sides is to remove/cleanup those 'rollout' knobs as soon as the first few rollouts are safe and cleared

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1329,6 +1335,21 @@ def _get_split_source(pg):
+    No op for other backends.
+    """
+    backend = None
+    try:
+        backend = pg._get_backend(torch.device("cuda"))
+    except RuntimeError:
+        pass
+    if is_nccl_available() and isinstance(backend, ProcessGroupNCCL):
```
This feels a bit heuristic/arbitrary.
If an `if-else` is truly needed, shall we put it upfront as much as possible than wrapping it in some function whose name sounds generic?

shuqiangzhang replied:
I think is_nccl_available() is not needed here as we are checking isinstance(backend, ProcessGroupNCCL) anyway. "cuda" is needed to check the backend used for GPU devices.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -89,13 +89,22 @@ static std::vector<std::string> TORCH_NCCL_WAIT_TIMEOUT_DUMP_MILSEC = {
 static std::vector<std::string> TORCH_NCCL_COORD_CHECK_MILSEC = {
     "TORCH_NCCL_COORD_CHECK_MILSEC"};
 
+// Whether to abort the communicators when users call destroy_process_group().
+// If yes, communicators will be aborted when destroy_process_group is called,
+// but not in destructor.
+static std::vector<std::string> TORCH_NCCL_ABORT_IN_DESTROY_PG = {
+    "TORCH_NCCL_ABORT_IN_DESTROY_PG"};
```
Same.

shuqiangzhang replied:
same explanation, we will remove this ENV after the safe rollout

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -89,13 +89,22 @@ static std::vector<std::string> TORCH_NCCL_WAIT_TIMEOUT_DUMP_MILSEC = {
 constexpr auto kProcessGroupNCCLDefaultTimeout =
     std::chrono::milliseconds(10 * 60 * 1000);
 
+constexpr auto kProcessGroupNCCLAbortTimeout =
+    std::chrono::milliseconds(10 * 60 * 1000);
+
```
I wonder if this can share one constant with the heartbeat monitor? The heartbeat monitor is also waiting on NCCL abort timeout.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1531,7 +1592,13 @@ void ProcessGroupNCCL::watchdogHandler() {
          /* no increment */) {
       auto& work = *it;
-      work.checkAndSetException();
+      // When terminateProcessGroup_ is true, communicators have already been
+      // aborted, So cannot check exception based on them. But watchdog needs to
```
This seems to apply to any method of WorkNCCL because any method can touch ncclComm_. Would be nice to figure out a holistic solution. For example, if `terminateProcessGroup_.load()` is true, can this thread just return?

shuqiangzhang replied:
Return was my first implementation, however, it has conflicts with some existing assumption that the thread would continue to run to finish already enqueued work checks, e.g., dec_pending_event_queries

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -557,7 +566,7 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // Provides an API to abort the ProcessGroup (similar to ncclCommAbort)
   // instead of relying on ProcessGroupNCCL destructor.
-  void abort(c10::optional<std::string> abortReason = c10::nullopt);
+  bool abort(c10::optional<std::string> abortReason = c10::nullopt);
```
nit: add a documentation regarding the bool?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -733,6 +742,12 @@ class TORCH_API ProcessGroupNCCL : public Backend {
       size_t timeout_sec = 30);
 
+  // A helper function to wait for a future to complete or timeout.
+  void waitForFutureOrTimeout(
+      std::future<bool>& fut,
+      const std::chrono::milliseconds& timeOutMilSec,
+      const std::string& futDescription);
+
```
nit: I wonder if this function needs to be a method of `ProcessGroupNCCL`? The declaration and implementation seem quite generic.

shuqiangzhang replied:
The only thing special to ProcessGroupNCCL is the exception info including the logPrefix making things easier to debug. with c++ 20 with coroutines, we should not need this helper.

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge

shuqiangzhang committed:
Update on "[c10d] explicitly abort communicators in destroy_process_group call"


Summary:
This PR tries to resolve issue #119215.

Basically,  processgroup shutdown (and hence ncclCommAbort) is called in
destroy_process_group APIs for the corresponding PGs. and in the
destructor of ProcessGroup, we avoid calling abort/ncclCommAbort.
Instead, it just checks if the users have explicitly already called destroy_process_group. If
not, Destructor will log a warning and encourage/expect users to do so
for cleanup of resources of PGs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1050,57 +1097,70 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
+  LOG(INFO) << logPrefix()
+            << "Launching ProcessGroupNCCL abort asynchrounously.";
+  std::future<bool> fut = std::async(std::launch::async, [this, abortReason]() {
+    return this->abort(abortReason);
+  });
 
-  workMetaListCV_.notify_one();
+  waitForFutureOrTimeout(fut, options_->timeout, "ProcessGroup abort");
```
One more thing:
If we had to use future on ncclCommAbort, we might as well just consider moving back to ncclCommDestroy.

Previously we prefer ncclCommAbort over ncclCommDestroy during regular teardown because it will not hang. Now that this characteristic is gone, and ncclCommAbort is not waitable (the only NCCL API not supporting nonblocking mode because it regards itself not possible to hang, lol), we can go back to ncclCommDestroy, but this time with nonblocking mode (i.e. in case ncclCommDestroy times out, kill it).

shuqiangzhang replied:
Good point, I also considered using nonblocking mode (but did not know that ncclCommAbort not supporting it and we need to change to ncclCommDestroy). Since nonblocking mode rollout has wider impact other than just abort() but also initialiation and every nccl call, we'd better have a separate rollout plan on that, during which we can try change to ncclCommDestroy

shuqiangzhang commented:
@pytorchbot merge
